core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceChangeStreamListener.java

```diff
@@ -0,0 +1,245 @@
+  private final Map<BsonValue, ChangeEvent<BsonDocument>> events;
+  private Thread streamThread;
+  private ReadWriteLock nsLock;
+  private final Deque<Callback<ChangeEvent<BsonDocument>, Object>> callbackQueue;
+  private Stream<ChangeEvent<BsonDocument>> currentStream;
+
+  NamespaceChangeStreamListener(
+      final MongoNamespace namespace,
+      final NamespaceSynchronizationConfig nsConfig,
+      final CoreStitchServiceClient service,
+      final NetworkMonitor networkMonitor,
+      final AuthMonitor authMonitor
+  ) {
+    this.namespace = namespace;
+    this.nsConfig = nsConfig;
+    this.service = service;
+    this.networkMonitor = networkMonitor;
+    this.authMonitor = authMonitor;
+    this.events = new HashMap<>();
+    this.nsLock = new ReentrantReadWriteLock();
+    this.logger =
+        Loggers.getLogger(
+            String.format("NamespaceChangeStreamListener-%s", namespace.toString()));
+    this.callbackQueue = new ArrayDeque<>();
+    this.networkMonitor.addNetworkStateListener(this);
+  }
+
+  @Override
+  public void onNetworkStateChanged() {
+    if (!this.networkMonitor.isConnected()) {
+      this.stop();
+      this.nsConfig.setStale();
+    } else {
+      this.start();
+    }
+  }
+
+  /**
+   * Opens the stream in a background thread.
+   */
+  public void start() {
+    nsLock.writeLock().lock();
+    try {
+      if (streamThread != null) {
+        return;
+      }
+      streamThread =
+          new Thread(new NamespaceChangeStreamRunner(
+              new WeakReference<>(this), networkMonitor, logger));
+      streamThread.start();
+    } finally {
+      nsLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Stops the background stream thread.
+   */
+  public void stop() {
+    nsLock.writeLock().lock();
+    try {
+      if (streamThread == null) {
+        return;
+      }
+
+      streamThread.interrupt();
+      try {
+        if (streamThread.isAlive()) {
+          close();
+          streamThread.join();
+        }
+      } catch (final Exception e) {
+        return;
+      }
+      streamThread = null;
+    } catch (Exception e) {
+      e.printStackTrace();
+    } finally {
+      nsLock.writeLock().unlock();
+    }
+  }
+
+  void queueWatcher(final Callback<ChangeEvent<BsonDocument>, Object> callback) {
+    callbackQueue.add(callback);
+  }
+
+  void clearWatchers() {
+    for (int i = 0; i < callbackQueue.size(); i++) {
+      callbackQueue.remove().onComplete(
+          OperationResult.<ChangeEvent<BsonDocument>, Object>failedResultOf(null));
+    }
+  }
+
+  void close() {
+    if (currentStream != null) {
+      try {
+        currentStream.close();
+      } catch (final IOException e) {
+        e.printStackTrace();
+      } finally {
+        currentStream = null;
+      }
+    }
+  }
+
+  /**
+   * Opens the next next.
+   */
+  void stream() {
+    logger.info("stream START");
+    if (!networkMonitor.isConnected()) {
+      logger.info("stream END - Network disconnected");
```

Comment:
[q] does this need to trigger onNetworkStateChanged or is that already being handled?